File: source/lib/native/frame.ts
// The binary input shape for ops that take whole files: a list of named byte
// strings, so a log's bytes cross the boundary as they are rather than as a
// JSON string escaped byte by byte.
//
// `[u32 count] ( [u32 name length][name] [u32 data length][data] )*`, little
// endian, mirroring `frame.rs`.

export type NamedBytes = {name: string; data: Uint8Array};

const encoder = new TextEncoder();

export const encodeFrame = (items: readonly NamedBytes[]): Uint8Array => {
	const names = items.map(item => encoder.encode(item.name));
	const total = items.reduce(
		(sum, item, index) => sum + 8 + names[index]!.length + item.data.length,
		4,
	);

	const out = new Uint8Array(total);
	const view = new DataView(out.buffer);
	let offset = 0;

	const u32 = (value: number) => {
		view.setUint32(offset, value, true);
		offset += 4;
	};

	u32(items.length);

	for (const [index, item] of items.entries()) {
		const name = names[index]!;
		u32(name.length);
		out.set(name, offset);
		offset += name.length;
		u32(item.data.length);
		out.set(item.data, offset);
		offset += item.data.length;
	}

	return out;
};

File: source/rust/epiq-core/src/frame.rs
//! The one binary input shape: a list of named byte strings, for ops that take
//! whole files. JSON would cost an escape pass over every byte of every log.
//!
//! `[u32 count] ( [u32 name length][name] [u32 data length][data] )*`, little
//! endian throughout.

pub struct NamedBytes<'a> {
    pub name: &'a str,
    pub data: &'a [u8],
}

pub fn decode(input: &[u8]) -> Result<Vec<NamedBytes<'_>>, String> {
    let mut cursor = 0usize;

    let take_u32 = |cursor: &mut usize| -> Result<usize, String> {
        let end = *cursor + 4;
        let bytes: [u8; 4] = input
            .get(*cursor..end)
            .ok_or("frame: truncated length")?
            .try_into()
            .expect("four bytes");
        *cursor = end;
        Ok(u32::from_le_bytes(bytes) as usize)
    };

    let count = take_u32(&mut cursor)?;
    let mut items = Vec::with_capacity(count);

    for _ in 0..count {
        let name_len = take_u32(&mut cursor)?;
        let name = input
            .get(cursor..cursor + name_len)
            .ok_or("frame: truncated name")?;
        cursor += name_len;

        let data_len = take_u32(&mut cursor)?;
        let data = input
            .get(cursor..cursor + data_len)
            .ok_or("frame: truncated data")?;
        cursor += data_len;

        items.push(NamedBytes {
            name: std::str::from_utf8(name).map_err(|e| format!("frame: name: {e}"))?,
            data,
        });
    }

    if cursor != input.len() {
        return Err("frame: trailing bytes".into());
    }

    Ok(items)
}

#[cfg(test)]
pub fn encode(items: &[(&str, &[u8])]) -> Vec<u8> {
    let mut out = Vec::new();
    out.extend((items.len() as u32).to_le_bytes());

    for (name, data) in items {
        out.extend((name.len() as u32).to_le_bytes());
        out.extend(name.as_bytes());
        out.extend((data.len() as u32).to_le_bytes());
        out.extend(*data);
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trips() {
        let framed = encode(&[("a.jsonl", b"one\n"), ("b.jsonl", b"")]);
        let items = decode(&framed).unwrap();

        assert_eq!(items.len(), 2);
        assert_eq!(items[0].name, "a.jsonl");
        assert_eq!(items[0].data, b"one\n");
        assert_eq!(items[1].data, b"");
    }

    #[test]
    fn refuses_a_short_frame() {
        assert!(decode(&[1, 0, 0, 0, 5, 0]).is_err());
        assert!(decode(&[0, 0, 0, 0, 9]).is_err());
    }
}
